Variuos fixes after some testing

emullaraj opened the pull request:
This pull request is to fix some bugs I found during testing of the SDK. I found there are still some bugs though, most of them because the response classes are missing the Xml root element in the constructor.

Thank you for this great work!

Looking forward to future collaborations.
em

davidtsadler commented:
Thank you for the pull request. Can I ask what issues this will resolve? I've looked at some of the changes and I'm having a hard time figuring out why they have been implement.

emullaraj commented:
Sorry, you are right :) 

* AddItemsResponseType.php
* AddItemsResponseType.php
* AddItemResponseType.php
* EndItemsResponseType.php
* ReviseItemResponseType.php
* FindItemsByProductResponse.php
* FindItemsByKeywordsResponse.php

Adding XmlRootElements to the contructor makes the toRequestXml() method generate a valid XML format.

* BaseType.php
During my tests I found out that values from Responses needed to be escaped, so I added this functionallity to Type\BaseType

* ProductIDCodeType.php
I added the ProductIDCodeType.php just to have it also under the Finding service.

* BaseFindingServiceResponse.php
The response of the Finding service contained an "Ack" parameter, so I added it in BaseFindingServiceResponse.php by suposing it is a common parameter for all responses.

I'll be happy to help if there is anything else that I should comment.

Regards,
Eno

davidtsadler commented:
Cheers for the extra info. I'm interested to know under what circumstances the Finding service was returning the Ack field. As far as I know this should be already handled in the ack field (Note the case)

emullaraj commented:
Allright, now I get it... I was looking for the Ack paramenter in my code. That "note the case" made me double check it :+1: 

So it is unnecesary adding the Ack in the BaseFindingServiceResponse.php, the ack in BaseServiceResponse.php is correct.

davidtsadler commented on src/DTS/eBaySDK/Trading/Types/AbstractRequestType.php:
```diff
@@ -48,7 +48,7 @@ class AbstractRequestType extends \DTS\eBaySDK\Types\BaseType
         ),
         'DetailLevel' => array(
-            'type' => 'string',
+            'type' => '\DTS\eBaySDK\Trading\Enums\DetailLevelCodeType[]',
```
I can see what you are trying to do here but it does not work. Both the following lines will throw an exception because of this change.

```php
$request->DetailLevel = 'ReturnAll';

$request->DetailLevel = Enums\DetailLevelCodeType::C_RETURN_ALL;
```

In addition I'm not sure if the SDK will be able to check that a constant has been passed to it. All the SDK will see is that a string value has been assigned.

davidtsadler commented on src/DTS/eBaySDK/Types/BaseType.php:
```diff
@@ -459,7 +569,8 @@ private static function ensurePropertyType($class, $name, $value)
 
-        if ($expectedType !== $actualType && 'array' !== $actualType) {
+        if (($expectedType !== $actualType && 'array' !== $actualType) 
+                && ('double' === $expectedType &&  !is_numeric($value))) {
```
This changes causes the tests to fail with the following errors.

DTS\eBaySDK\Types\Test\ConstructTest::testSettingInvalidPropertyTypeViaCtor
Failed asserting that exception of type "\DTS\eBaySDK\Exceptions\InvalidPropertyTypeException" is thrown.

DTS\eBaySDK\Types\Test\SimpleClassTest::testSettingPropertyWithAnInvalidType
Failed asserting that exception of type "\DTS\eBaySDK\Exceptions\InvalidPropertyTypeException" is thrown.

davidtsadler commented on src/DTS/eBaySDK/Types/BaseType.php:
```diff
@@ -553,12 +727,27 @@ private static function encodeValueXml($value)
         else if (is_bool($value)){
             return $value ? 'true' : 'false';
+        }
+        else if (is_string($value)){
```
This causes the tests to fail with the following errors.

DTS\eBaySDK\Types\Test\AmountClassTest::testToXml
Failed asserting that two DOM documents are equal.

DTS\eBaySDK\Types\Test\ComplexClassTest::testToXml
Failed asserting that two DOM documents are equal.

davidtsadler commented:
Thank you for the pull request and many apologies for not getting back to this sooner. I'm currently busy getting the 1.0 release of the SDK out and I've let things slip a little. Unfortunately I can't accept the pull request as it stands. There are many reasons for this, which I will go into below, but the main reason is that some of the changes cause the unit tests to fail. The tests can be run with the phpunit command `vendor/bin/phpunit` or the `make test` command. 

I've added notes to some of the [changed files](https://github.com/davidtsadler/ebay-sdk-php/pull/10/files) to highlight some of the issues that I have found. Below are a few comments on some of the features that the pull request would add.

### XML

I've noticed that you have made some changes so that it will be possible to obtain the XML of a response object.

```php
$response = $service->addItems($request);

echo $response->toRequestXml();
```

This feature will be something that I plan to add to the 1.0 release of the SDK. This will be implemented via a new `toXml` method as the current `toRequestXml` method sees wrongly named when calling it on a response object.

```php
$response = $service->addItems($request);

echo $response->toXml();
```

I need to draw you attention to the changes where you are updating the [$requestXmlRootElementNames](https://github.com/davidtsadler/ebay-sdk-php/pull/10/files#diff-d6e6c2281af769e72b4ba60c68a6a01bR59) property in various response classes. If you look at the comments in the top of these files you will see the following.

```php
/**
 * THE CODE IN THIS FILE WAS GENERATED FROM THE EBAY WSDL USING THE PROJECT:
 *
 * https://github.com/davidtsadler/ebay-api-sdk-php
 *
 ``` 

It's not well known but most of the code that makes up the SDK is generated automatically and not written manually. The PHP files found in the various [types](https://github.com/davidtsadler/ebay-sdk-php/tree/master/src/DTS/eBaySDK/Trading/Types) and [enums](https://github.com/davidtsadler/ebay-sdk-php/tree/master/src/DTS/eBaySDK/Finding/Enums) directories where created by parsing the eBay WSDLs using a [project](https://github.com/davidtsadler/ebay-api-sdk-php) that is independent of the SDK. The reason for the automation is that I can just *re-build* the SDK whenever eBay release a new version of their API. I have a [script](https://github.com/davidtsadler/ebay-api-sdk-php/blob/master/Makefile#L33) that downloads the latest WSDLs which are used to build the updated PHP files. Sadly this means that **any** code that is added manually to these files is lost the next time the SDK is updated. To implement features that require changes to these types of files, such as the `toRequestXml` method, requires the actual build process to be updated. In this instance I will be adding this update as part of implementing the new `toXml` method in the 1.0 release. 

### JSON

I can see that you have added a `toRequestJSON` method. This is also a new feature that I will be adding in the 1.0 release. It will be implemented in the appropriately named method `toJSON`. Sadly as it stands your `toRequestJSON` method does not work.

```PHP
$request = new Types\FindItemsByProductRequest();
$productId = new Types\ProductId();
$productId->value = '085392246724';
$productId->type = 'UPC';
$request->productId = $productId;
$itemFilter = new Types\ItemFilter();
$itemFilter->name = 'ListingType';
$itemFilter->value[] = 'Auction';
$itemFilter->value[] = 'AuctionWithBIN';
$request->itemFilter[] = $itemFilter;
$request->itemFilter[] = new Types\ItemFilter(array(
    'name' => 'MinPrice',
    'value' => array('1.00')
));
$request->itemFilter[] = new Types\ItemFilter(array(
    'name' => 'MaxPrice',
    'value' => array('10.00')
));
$request->sortOrder = 'CurrentPriceHighest';
$request->paginationInput = new Types\PaginationInput();
$request->paginationInput->entriesPerPage = 10;
$request->paginationInput->pageNumber = 1;

echo $request->toRequestJSON();
```

The ouput from `toRequestJSON` does not include all the `itemFilter` entries and their values, and incorrectly adds " characters to some of the values.

```JSON
{
    "findItemsByProductRequest": {
        "paginationInput": {
            "pageNumber": 1,
            "entriesPerPage": 10
        },
        "sortOrder": "CurrentPriceHighest",
        "productId": {
            "type": "UPC",
            "value": "\"085392246724\""
        },
        "itemFilter": {
            "name": "MaxPrice",
            "value": "10.00"
        }
    }
}
```

The correct JSON should be.

```JSON
{
    "findItemsByProductRequest": {
        "paginationInput": {
            "pageNumber": 1,
            "entriesPerPage": 10
        },
        "sortOrder": "CurrentPriceHighest",
        "productId": {
            "value": "085392246724",
            "type": "UPC"
        },
        "itemFilter": [
            {
                "name": "ListingType",
                "value": [
                    "Auction",
                    "AuctionWithBIN"
                ]
            },
            {
                "name": "MinPrice",
                "value": [
                    "1.00"
                ]
            },
            {
                "name": "MaxPrice",
                "value": [
                    "10.00"
                ]
            }
        ]
    }
}
```

### Encoding HTML entities

You have added code to encode various characters into HTML entities when used in the XML. This is something that I have mentioned in the [Version 1.0.0 Discussion](https://github.com/davidtsadler/ebay-sdk-php/issues/3) and will be implemented as part of the 1.0 release.

### New ProductIDCodeType enum

I have to admit that I'm in two minds about the addition of the [ProductIDCodeType](https://github.com/davidtsadler/ebay-sdk-php/pull/10/files#diff-90149b8b67c40db7bf9e646141222422R22) enum. On the one hand I can see it been useful but on the other I'm not sure about the implementation. The actual file is located in a directory that is overwritten during the automatic rebuild of the SDK and so will be lost. I can make the changes to prevent this but I will need time to do this and it may not be the top priority during the 1.0 development.

### Apologies

I must apologise at this point as it looks like I'm wasting other developer's time. One of the improvements that I want to implement for the 1.0 release is been more open about what new features are been worked on. This will be done by starting regular blog posts and tweets that should keep people informed. This way I hope to prevent developers from working on duplicate features and to also provide people with a way of giving feedback on how the SDK can be improved. I also plan to improve the SDK's documentation and provide more information that is currently not that well know.

Again I apologise that I won't be able to accept this pull request.
